### PR TITLE
HWY-217: Add `needs_validation` method to `ConsensusValueT`.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/HWY-217

If consensus value is empty then consensus doesn't have to ask for its validation and add the vertex to the state immediately, saving some reactor events and processing.

It also simplifies tests since we can propose empty vectors without having to handle `ResolveValidity` requests.

### CODE REVIEW

File: node/src/components/consensus/traits.rs [hunk 1]
```diff
@@ -19,8 +19,9 @@ impl<VID> ValidatorIdT for VID where VID: Eq + Ord + Clone + Debug + Hash {}
 pub(crate) trait ConsensusValueT:
     Eq + Clone + Debug + Hash + Serialize + DeserializeOwned
 {
+    /// Returns whether the consensus value is empty.
+    fn is_empty(&self) -> bool;
```
Comment (afck) [inline node/src/components/consensus/traits.rs:23]: I'd prefer to call it something like `needs_validation`, so the trait doesn't care about whether these are lists or something else. And it would make it more natural to e.g. set it to always `false` for some tests, or in case requirements in production change.

File: node/src/components/consensus/protocols/highway.rs [hunk 2]
```diff
@@ -299,14 +299,22 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
...
-                                    .push(vv);
-                                results.push(ProtocolOutcome::ValidateConsensusValue(
-                                    sender, value, timestamp,
-                                ));
+                                // It's a block…
+                                if value.is_empty() {
+                                    // Consensus value is empty. Nothing to validate.
+                                    let now = Timestamp::now();
+                                    results.extend(self.add_valid_vertex(vv, rng, now));
+                                    state_changed = true;
```
Comment (afck) [inline node/src/components/consensus/protocols/highway.rs:307]: Can this be deduplicated and merged with the `else` below?
```rust
match (vertex.value().cloned(), vertex.timestamp()) {
    (Some(value), Some(timestamp)) if !value.is_empty() => {
        // Request validation before adding it to the state.
        // …
    }
    (_, _) => {
        // No validation required. Add to the state.
        // …
    }
}
```
Or does the borrow checker complain about that?

**Review by afck — APPROVED** [review]
I'm a bit on the fence, because if in doubt I'd always prioritize simplicity of production code over performance and simplicity of the test code. But if it makes a big difference for the tests I'm fine with it.

**Comment by goral09** [comment]
> I'm a bit on the fence, because if in doubt I'd always prioritize simplicity of production code over performance and simplicity of the test code. But if it makes a big difference for the tests I'm fine with it.

I haven't yet really seen that much simplification of the tests, some but not huge, but I think it's non-intrusive change that makes sense IMO.

File: node/src/components/consensus/traits.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/traits.rs:23]: Good idea.

File: node/src/components/consensus/protocols/highway.rs (hunk 2 above)
Comment (goral09) [reply node/src/components/consensus/protocols/highway.rs:307]: I haven't really tried. I did see the duplication but I thought I will go with the simplicity for the reader's sake.

Comment (goral09) [reply node/src/components/consensus/protocols/highway.rs]: Done in [`09ae779` (#653)](https://github.com/CasperLabs/casper-node/pull/653/commits/09ae779c37005427481d73af152128e242604999)

File: node/src/components/consensus/traits.rs (hunk 1 above)
Comment (goral09) [reply node/src/components/consensus/traits.rs]: Done in [`06fe993` (#653)](https://github.com/CasperLabs/casper-node/pull/653/commits/06fe99365ed796bbdf1d1d3ce0528cac422999cd)

**Comment by goral09** [comment]
bors merge
